### PR TITLE
Fix Errrors And release New Version

### PR DESCRIPTION
**Pull request by mohammad-malekpaeen**
Please code review and merge

### CODE REVIEW

File: readme.txt [hunk 1]
```diff
@@ -33,6 +33,14 @@ Thank you so much for using IDPay Payment Gateway.
...
 == Changelog ==
 
+= 2.2.0, June 13, 2022 =
+* Fix encode and sanitize config fileds
+* Improve Un-Used Params And Functions
+* Fix Correct Redirect When Strange errors
+* Improve Security And Stripe URI In Payment Flow
+* Change Structure Database
+* Update Style When Show Result 
+
```
Comment (meysamrazmi) [inline readme.txt]:
```suggestion
= 2.2.0, June 13, 2022 =
* Improve sanitizing
* Change redirecting behavior after payment

```

File: readme.txt [hunk 2]
```diff
@@ -1,7 +1,7 @@
 === IDPay for Contact Form 7 ===
-Contributors: JMDMahdi, imikiani, meysamrazmi, vispa
-Tags: IDPay, contact form 7, form, payment, contact form
-Stable tag: 2.1.4
+Contributors: JMDMahdi, imikiani, meysamrazmi, vispa, Mohammad Malek(MimDeveloper.Tv)
+Tags: IDPay, Contact Form 7, Form, Payment, Contact Form
+Stable tag: 2.2.0
 Tested up to: 5.7
```
Comment (meysamrazmi) [inline readme.txt]:
```suggestion
Tested up to: 6.0
```

File: includes/Plugin.php [hunk 3]
```diff
@@ -100,32 +118,33 @@ function goBack() {
...
-    public static function update() {
+    public static function update()
+    {
         global $wpdb;
         $table_name = $wpdb->prefix . "cf7_transactions";
-        $version = get_option( 'idpay_cf7_version', '1.0' );
+        $version = get_option('idpay_cf7_version', '1.0');
 
-        if ( version_compare( $version, '2.1.1' ) < 0 ) {
+        if (version_compare($version, '2.1.1') < 0) {
```
Comment (meysamrazmi) [inline includes/Plugin.php:141]: you didn't add another else so you won't create your new table for those who are already using the plugin

@MimDeveloperTv

File: includes/Plugin.php [hunk 4]
```diff
@@ -45,17 +48,29 @@ public static function activate() {
                 PRIMARY KEY id (id)
             );";
             require_once(ABSPATH . 'wp-admin/includes/upgrade.php');
-            dbDelta( $sql );
+            dbDelta($sql);
+
+            if ($wpdb->get_var("show tables like '$callback_table_name'") != $callback_table_name) {
+                $sql = "CREATE TABLE $callback_table_name (
+               id bigint(11) NOT NULL AUTO_INCREMENT,
```
Comment (meysamrazmi) [inline includes/Plugin.php]: lint problem

File: includes/Payment/Result.php [hunk 5]
```diff
@@ -35,9 +35,10 @@ public function register() {
      * @return string
      */
     public function handler( $atts ) {
-        if( !empty( $_GET['status'] ) && !empty( $_GET['message'] ) ){
-            $color = $_GET['status'] == 'failed' ? '#f44336' : '#8BC34A';
-            return '<b style="color:'. $color .';text-align:center;display: block;">' . $_GET['message'] . '</b>';
+        if(!empty( $_GET['order_id'])) {
+            require_once( dirname(__DIR__) . '/Functions.php' );
+            global $wpdb;
+            return fetch_callback_response($wpdb, $_GET['order_id']);
```
Comment (meysamrazmi) [inline includes/Payment/Result.php]: You don't need to pass the db variable to the function

File: includes/Payment/Result.php [hunk 6]
```diff
@@ -35,9 +35,10 @@ public function register() {
      * @return string
      */
     public function handler( $atts ) {
-        if( !empty( $_GET['status'] ) && !empty( $_GET['message'] ) ){
-            $color = $_GET['status'] == 'failed' ? '#f44336' : '#8BC34A';
-            return '<b style="color:'. $color .';text-align:center;display: block;">' . $_GET['message'] . '</b>';
+        if(!empty( $_GET['order_id'])) {
```
Comment (meysamrazmi) [inline includes/Payment/Result.php]: use a more accurate variable name for this plugin. `order_id` is so general

File: includes/Functions.php [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            'id' => $order_id,
+            'response' => json_encode([
+                'trans_id'=>$trans_id,
+                'track_id'=>$track_id,
+                'status'=>$status
+            ]),
+            'message' => json_encode($message),
+            'created_at' => time(),
+        ];
+        $db->insert( $db->prefix . $tableName, $row, array('%s','%s','%s','%s') );
```
Comment (meysamrazmi) [inline includes/Functions.php]: you didn't care about linting in every place

File: includes/Functions.php [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+/**
+ * @param $db
+ * @param $order_id
+ * @param $trans_id
+ * @param $track_id
+ * @param $status
+ * @param $message
+ * @return void
+ */
+function create_callback_response($db, $order_id, $trans_id, $track_id, $status, $message)
```
Comment (meysamrazmi) [inline includes/Functions.php]: no need to pass `$db` to this function. this is a global variable

File: includes/Functions.php [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+{    $tableName = 'cf7_callbacks';
```
Comment (meysamrazmi) [inline includes/Functions.php]: WTF?

File: includes/Callback.php [hunk 10]
```diff
@@ -175,10 +153,19 @@ function call_gateway_endpoint( $url, $args ) {
...
         exit();
     }
 } else {
-    wp_redirect( add_query_arg( ['status' => 'failed', 'message' => __( 'Transaction not found', 'idpay-contact-form-7' )], $value['return'] ) );
+
+    $status = 'failed' ;
+    $message  = __( 'Transaction not found', 'idpay-contact-form-7' ) ;
+    create_callback_response( $wpdb, time(), null,null,$status,$message);
+
+    wp_redirect( add_query_arg([], $value['return'] ) );
```
Comment (meysamrazmi) [inline includes/Callback.php]: why you didn't set the order_id in query argumants

File: includes/Functions.php [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    $row = $db->get_row( $db->prepare( "SELECT * FROM " . $db->prefix . $tableName ." WHERE id='%s'", $order_id ) );
+    if ( $row == NULL ) {
+        $row = [
+            'id' => $order_id,
+            'response' => json_encode([
+                'trans_id'=>$trans_id,
+                'track_id'=>$track_id,
+                'status'=>$status
+            ]),
```
Comment (meysamrazmi) [inline includes/Functions.php]: there is no need to save these ids in this table. this table is being used just for redirecting

File: includes/Functions.php [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+            'message' => json_encode($message),
```
Comment (meysamrazmi) [inline includes/Functions.php]: `$message` is already a string. why do you encode it here?

**Review by meysamrazmi — COMMENTED** [review]
final

File: includes/Callback.php (hunk 10 above)
Comment (mohammad-malekpaeen) [reply includes/Callback.php]: ok

File: includes/Functions.php (hunk 8 above)
Comment (mohammad-malekpaeen) [reply includes/Functions.php]: ok

File: includes/Callback.php [hunk 13]
```diff
@@ -1,184 +1,165 @@
...
                 '%s',
             ),
-            array( '%d' )
+            array('%d')
         );
-        wp_redirect( add_query_arg( ['status' => 'failed', 'message' => filled_message( $value['failed_message'], $track_id, $order_id )], $value['return'] ) );
+
+        $status = 'failed';
+        $message = filled_message($value['failed_message'], $track_id, $order_id);
+        create_callback_response($order_id,$status, $message);
```
Comment (meysamrazmi) [inline includes/Callback.php]: `$order_id,$status`
lint

File: includes/Callback.php [hunk 14]
```diff
@@ -1,184 +1,165 @@
...
+    if ($row !== NULL) {
+        if ($row->status == 'completed') {
 
-    $row = $wpdb->get_row( $wpdb->prepare( "SELECT * FROM " . $wpdb->prefix . "cf7_transactions WHERE trans_id='%s'", $id ) );
-    if ( $row !== NULL ) {
-        if ( $row->status == 'completed' ) {
-            wp_redirect( add_query_arg( ['status' => 'success', 'message' => filled_message( $value['success_message'], $row->track_id, $row->order_id )], $value['return'] ) );
+            $status = 'success';
+            $message = filled_message($value['success_message'], $row->track_id, $row->order_id);
+            create_callback_response($wpdb, $order_id, $trans_id, $track_id, $status, $message);
```
Comment (meysamrazmi) [inline includes/Callback.php]: you didn't pass correct parameters to the function

File: includes/Callback.php [hunk 15]
```diff
@@ -1,184 +1,165 @@
...
 
-    $response = call_gateway_endpoint( 'https://api.idpay.ir/v1.1/payment/verify', $args );
+    $response = call_gateway_endpoint('https://api.idpay.ir/v1.1/payment/verify', $args);
+    if (is_wp_error($response)) {
 
-    if ( is_wp_error( $response ) ) {
-        wp_redirect( add_query_arg( ['status' => 'failed', 'message' => $response->get_error_message()], $value['return'] ) );
+        $status = 'failed';
+        $message = $response->get_error_message();
+        create_callback_response($order_id,$status, $message);
```
Comment (meysamrazmi) [inline includes/Callback.php]: lint

File: includes/Functions.php [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+ */
+function create_callback_response($order_id, $status, $message)
+{
+    global $wpdb;
+    $tableName = 'cf7_callbacks';
+    $row = $wpdb->get_row($wpdb->prepare("SELECT * FROM " . $wpdb->prefix . $tableName . " WHERE id='%s'", $order_id));
+    if ($row == NULL) {
+        $row = [
+            'id' => $order_id,
+            'response' => json_encode(['status' => $status]),
```
Comment (meysamrazmi) [inline includes/Functions.php]: I said you should change the response to status. it's not necessary to save it as json

File: includes/Payment/Payment.php [hunk 17]
```diff
@@ -17,170 +17,172 @@
...
+
+        if (empty($api_key)) {
+            wp_redirect(add_query_arg('idpay_error', __('IDPay should be configured properly', 'idpay-contact-form-7')));
             exit;
         }
 
-		if ( empty( $amount ) ) {
-            wp_redirect( add_query_arg( 'idpay_error', __( 'Amount can not be empty', 'idpay-contact-form-7' )) );
+        if (empty($amount)) {
+            wp_redirect(add_query_arg('idpay_error', __('Amount can not be empty', 'idpay-contact-form-7')));
```
Comment (meysamrazmi) [inline includes/Payment/Payment.php]: these redirects should have been changed too

File: includes/Payment/Payment.php [hunk 18]
```diff
@@ -17,170 +17,172 @@
...
+            'timeout' => 15,
+        );
+
+        $response = $this->call_gateway_endpoint('https://api.idpay.ir/v1.1/payment', $args);
+        if (is_wp_error($response)) {
+            $error = $response->get_error_message();
+            $row['status'] = 'failed';
+            $row['log'] = $error;
+            $wpdb->insert($wpdb->prefix . "cf7_transactions", $row, $row_format);
+            wp_redirect(add_query_arg('idpay_error', $error));
```
Comment (meysamrazmi) [inline includes/Payment/Payment.php]: this one too

File: includes/Payment/Payment.php [hunk 19]
```diff
@@ -17,170 +17,172 @@
...
+        $http_status = wp_remote_retrieve_response_code($response);
+        $result = wp_remote_retrieve_body($response);
+        $result = json_decode($result);
+
+        if ($http_status != 201 || empty($result) || empty($result->id) || empty($result->link)) {
+            $error = sprintf('Error : %s (error code: %s)', $result->error_message, $result->error_code);
+            $row['status'] = 'failed';
+            $row['log'] = $error;
+            $wpdb->insert($wpdb->prefix . "cf7_transactions", $row, $row_format);
+            wp_redirect(add_query_arg('idpay_error', $error));
```
Comment (meysamrazmi) [inline includes/Payment/Payment.php]: this one too

File: includes/Payment/Payment.php [hunk 20]
```diff
@@ -17,170 +17,172 @@
...
+            $response = wp_safe_remote_post($url, $args);
+            if (is_wp_error($response)) {
+                $number_of_connection_tries--;
+                continue;
+            } else {
+                break;
+            }
+        }
+        return $response;
+    }
```
Comment (meysamrazmi) [inline includes/Payment/Payment.php]: use the functions.php instead

File: includes/Payment/Result.php [hunk 21]
```diff
@@ -34,11 +36,12 @@ public function register() {
      *
      * @return string
      */
-    public function handler( $atts ) {
-        if( !empty( $_GET['status'] ) && !empty( $_GET['message'] ) ){
-            $color = $_GET['status'] == 'failed' ? '#f44336' : '#8BC34A';
-            return '<b style="color:'. $color .';text-align:center;display: block;">' . $_GET['message'] . '</b>';
+    public function handler($atts)
+    {
+        if (!empty($_GET['order_id'])) {
```
Comment (meysamrazmi) [inline includes/Payment/Result.php]: as I said you should use a more specific term instead of `order_id`

File: includes/Plugin.php [hunk 22]
```diff
@@ -144,52 +164,64 @@ public static function update() {
...
+                    $wpdb->update($wpdb->prefix . 'postmeta',
+                        array('meta_value' => $meta_value),
+                        array('meta_id' => $row->meta_id),
+                        array('%s'),
+                        array('%d')
                     );
                 }
             }
         }
+        if(version_compare($version, '2.2.0') < 0)
```
Comment (meysamrazmi) [inline includes/Plugin.php]: Did you even test it???
the version you are checking should be the previous one

File: readme.txt [hunk 23]
```diff
@@ -33,6 +33,10 @@ Thank you so much for using IDPay Payment Gateway.
 
 == Changelog ==
 
+= 2.2.0, June 13, 2022 =
+	* Improve sanitizing
+	* Change redirecting behavior after payment
+
```
Comment (meysamrazmi) [inline readme.txt]: there is extra space before the line

**Review by meysamrazmi — DISMISSED** [review]
@MimDeveloperTv

File: includes/Callback.php (hunk 14 above)
Comment (mohammad-malekpaeen) [reply includes/Callback.php]: ok

File: includes/Callback.php (hunk 13 above)
Comment (mohammad-malekpaeen) [reply includes/Callback.php]: ok

File: includes/Callback.php (hunk 15 above)
Comment (mohammad-malekpaeen) [reply includes/Callback.php]: ok

File: includes/Functions.php (hunk 16 above)
Comment (mohammad-malekpaeen) [reply includes/Functions.php]: ok

File: includes/Payment/Payment.php (hunk 17 above)
Comment (mohammad-malekpaeen) [reply includes/Payment/Payment.php]: ok

**Review by mohammad-malekpaeen — COMMENTED** [review]
ok tanks

**Pull request merged**